camera_manager.py

```diff
@@ -1,14 +1,200 @@
+        self.source = self.cs.getVideo(camera=self.camera)
+        self.sink = self.cs.putVideo("Driver_Stream", FRAME_WIDTH, FRAME_HEIGHT)
+        # Width and Height are reversed here because the order of putVideo's width and height
+        # parameters are the opposite of numpy's (technically it is an array, not an actual image).
+        self.frame = np.zeros(shape=(FRAME_HEIGHT, FRAME_WIDTH, 3), dtype=np.uint8)
+
+    def get_frame(self) -> Tuple[float, np.ndarray]:
+        """Gets a frame from the camera.
+        Returns:
+            Frame_time, or 0 on error.
+            A numpy array of the frame, dtype=np.uint8, BGR.
+        """
+        frame_time, self.frame = self.source.grabFrameNoTimeout(image=self.frame)
+        return frame_time, self.frame
+
+    def send_frame(self, frame: np.ndarray) -> None:
+        """Sends a frame to the driver display.
+        Args:
+            frame: A numpy array image. (Should always be the same size)
+        """
+        self.sink.putFrame(frame)
+
+    def get_error(self) -> str:
+        """Gets an error from the camera.
+        Should be run by Vision when frame_time is 0.
+        Returns:
+            A string containing the camera's error.
+        """
+        return self.source.getError()
+
+    def notify_error(self, error: str) -> None:
+        """Sends an error to the console and the sink.
+        Args:
+            error: The string to send. Should be gotten by get_error().
+        """
+        print(error, file=sys.stderr)
+        self.sink.notifyError(error)
+
+    def set_camera_property(self, property, value) -> None:
+        self.camera.getProperty(property).set(value)
+
+
+class MockImageManager:
+    def __init__(self, image: np.ndarray, display_output: bool = False) -> None:
+        """Initialises a Mock Image Manager
+        Args:
+            image: A BGR numpy image array
+        """
+        self.image = image
+        self.display_output = display_output
+
+    def change_image(self, new_image: np.ndarray) -> None:
+        """Changes self.image.
+        Args:
+            new_image: The new image to switch to. Should be a numpy image array.
+        """
+        self.image = new_image
+
+    def get_frame(self) -> Tuple[float, np.ndarray]:
+        """Returns self.image.
+        Returns:
+            0.1: Simulates the frame_time
+            self.image, a BGR numpy array.
+        """
+        return 0.1, self.image.copy()
+
+    def send_frame(self, frame: np.ndarray):
+        if self.display_output:
+            cv2.imshow("Image", frame)
+            cv2.waitKey(0)
+
+    def get_error(self) -> str:
+        return "Error"
+
+    def notify_error(self, error: str) -> None:
+        """Prints an error to the console.
+        Args:
+            error: The string to print.
+        """
+        print(error, file=sys.stderr)
+
+    def set_camera_property(self, property, value) -> None:
+        ...
```

Comment:
`...` indicates to me that the method should be implemented in future (but isn't now and doing nothing in the time being is okay). Will these be implemented, or should they just be `pass`?

Reply:
Good catch - we want this code to actually be runnable, even if it doesn't do what it needs to.

I agree with Davo - these should be `pass`